pep-0394.txt

```diff
@@ -5,104 +5,181 @@ Last-Modified: $Date$
-  ``python`` on their shebang line.
-* When packaging software that is source compatible with both versions,
-  distributions may change such ``python`` shebangs to ``python3``.
-  This ensures software is used with the latest version of
-  Python available, and it can remove a dependency on Python 2.
+* The Python 3.x ``idle``, ``pydoc``, and ``python-config`` commands should
+  likewise be available as ``idle3``, ``pydoc3``, and ``python3-config``;
+  Python 2.x versions as ``idle2``, ``pydoc2``, and ``python2-config``.
+  The original commands should either invoke the same version of Python
+  as the ``python`` command, or not be available at all.
+* When packaging software, distributors are encouraged to change less specific
+  shebangs to more specific ones. This ensures software is used with the latest
+  version of Python available, and it can remove a dependency on Python 2.
+  The specifics are left on the distributors.
+  Examples of this include changing ``python`` shebangs to ``python3``
+  when Python 3.x is supported, changing ``python`` shebangs to ``python2``
+  when Python 3.x is not yet supported, or changing ``python3`` shebangs to
+  ``python3.8`` if the software is built with Python 3.8.
+* When a virtual environment (created by the PEP 405 ``venv`` package or a
+  similar tool such as ``virtualenv`` or ``conda``) is active, the ``python``
+  command should refer to the virtual environment's interpreter and should
+  always be available.
+  The ``python3`` or ``python2`` command (according to the environment's
+  interpreter) should also be available.
+
+For developers
+--------------
+
 * When reinvoking the interpreter from a Python script, querying
   ``sys.executable`` to avoid hardcoded assumptions regarding the
   interpreter location remains the preferred approach.
-* In controlled environments aimed at expert users, where being explicit
-  is valued over user experience (for example, in test environments and
-  package build systems), distributions may choose to not provide the
-  ``python`` command even if ``python2`` is available.
-  (All software in such a controlled environment must use ``python3`` or
-  ``python2`` rather than ``python``, which means scripts that deliberately
-  use ``python`` need to be modified for such environments.)
-* When a virtual environment (created by the PEP 405 ``venv`` package or a
-  similar tool) is active, the ``python`` command should refer to the
-  virtual environment's interpreter. In other words, activating a virtual
-  environment counts as deliberate user action to change the default
-  ``python`` interpreter.
+* While far from being universally available, ``python`` remains the
+  preferred spelling for explicitly invoking Python, as this is the
+  spelling that virtual environments make consistently available
+  across different platforms and Python installations.
+* In shebang lines, the preferred spelling is ``/usr/bin/env python``,
```

Comment:
Note that we proposed this PR without this recommendation so I leave the decision to you and @ncoghlan. I'd happily keep it or revert it back as you see fit.

Reply:
This case should be combined with the case on Line 104.

Reply:
I want to say that a general `/usr/bin/env python` recommendation outside of venvs is very dangerous.  When I was working on Debian/Ubuntu, such things lead to mysterious problems with serious ramifications.  So we went about actively discouraging such shebangs.

I think we need help from tooling here.  Ideally, end users wouldn't have to worry about it at all, and either the platform builds would modify the shebang (as I *think* but don't really remember any more whether Debuntu does this), or installing into a venv would rewrite the shebang.

So I'm quite concerned that a general recommendation will cause users serious problems.  If we want to make progress on this update and can't agree on shebang recommendations, we should remove it from this update and discuss further in a separate PR, bug, or forum.

Reply:
@warsaw Is there a difference here between system scripts and those distributed by third parties? If Debian ships a tool that happens to be written in Python (eg /usr/bin/calibre), it makes sense for its shebang to avoid /usr/bin/env, but if I'm going to give people a script that depends on Python, it makes more sense to let their environment select it.

Reply:
@Rosuav That's exactly the situation that broke people on Ubuntu.  They'd end up using some random Python they forgot they installed, with an incompatible library pip installed and it broke their script.  Then they blamed the wrong people, when really they should have pointed their script to the right Python executable, or fixed their `$PATH` correctly.  Okay, sure this is user error, but I know with a PEP 394 recommendation to use `/usr/bin/env`, it'll happen more often.  It might even be that they *think* they're running in a venv, but really they aren't.

Reply:
Note that the original PEP nor our PR have ever gone into `/usr/bin/python...` vs `/usr/bin/env python...` shebangs. What if we keep it that way and than we can argue about that in another PR?

Reply:
+1 - This is the only controversial bit still left IIRC.  I'm happy to defer this and push out an update with the bits we do agree with.  @encukou @ncoghlan Would that be okay with you?

Reply:
Yes!

Reply:
Resolved by removing the "preferred" wording, and instead qualifying this bullet point as being specific to scripts that are only expected to be run in virtual environments: https://github.com/python/peps/commit/4b20f8c130df2c6b9b1a9900ea41eca51999ee05

I then removed the two paragraphs about Py2 vs Py3, since they weren't needed any more, and the text flows better without them.